compiler/rustc_span/src/lib.rs

```diff
@@ -1044,17 +1044,26 @@ impl fmt::Debug for Span {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         // Use the global `SourceMap` to print the span. If that's not
         // available, fall back to printing the raw values.
-        with_session_globals(|session_globals| {
-            if let Some(source_map) = &*session_globals.source_map.borrow() {
-                write!(f, "{} ({:?})", source_map.span_to_diagnostic_string(*self), self.ctxt())
-            } else {
-                f.debug_struct("Span")
-                    .field("lo", &self.lo())
-                    .field("hi", &self.hi())
-                    .field("ctxt", &self.ctxt())
-                    .finish()
-            }
-        })
+
+        fn fallback(span: Span, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_struct("Span")
+                .field("lo", &span.lo())
+                .field("hi", &span.hi())
+                .field("ctxt", &span.ctxt())
+                .finish()
+        }
+        
+        if SESSION_GLOBALS.is_set() {
+            with_session_globals(|session_globals| {
+                if let Some(source_map) = &*session_globals.source_map.borrow() {
```

Comment:
This could be a let chain with a single fallback :)

Reply:
How do you propose to handle the `with_session_globals` scope using a let chain? I wasn't trying let chains previously, but either way I couldn't think of a nice one, since we have to take the borrowed data and return a Result while _within_ the closure.

For example, this would work, but I don't think it's an improvement; I think the control flow is much less clear to the reader since it is so combinator-ful (and my other attempts needed explicit success-or-failure variables):

```rust
if SESSION_GLOBALS.is_set()
    && let Some(_) = with_session_globals(|session_globals| {
        session_globals.source_map.borrow().as_ref().map(|source_map|
            write!(f, "{} ({:?})", source_map.span_to_diagnostic_string(*self), self.ctxt())
        )
    }).transpose()?
{
    Ok(())
} else {
    f.debug_struct("Span")
        .field("lo", &self.lo())
        .field("hi", &self.hi())
        .field("ctxt", &self.ctxt())
        .finish()
}
```

Reply:
oh, right, I somehow missed the closure when thinking about this